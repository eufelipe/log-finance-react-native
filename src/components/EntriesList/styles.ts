import styled from 'styled-components/native';

export const FlatList = styled.FlatList``;
