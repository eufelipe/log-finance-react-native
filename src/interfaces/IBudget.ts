import {Category} from 'models';

export default interface IBudget {
  description?: string;
  category: Category;
  value: number;
}
