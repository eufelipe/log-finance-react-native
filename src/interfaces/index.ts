export {default as ICategory} from './ICategory';
export {default as IEntry} from './IEntry';
export {default as IBudget} from './IBudget';
