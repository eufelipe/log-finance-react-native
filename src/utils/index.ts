export {default as number} from './number';
export {default as strings} from './strings';
