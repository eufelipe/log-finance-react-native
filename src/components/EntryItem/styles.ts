import styled from 'styled-components/native';
import {Text, TouchableOpacity} from 'styles/layout';

export const Container = styled(TouchableOpacity)`
  background: ${({theme}) => theme.colors.white};
  margin: ${({theme}) => theme.spacing.small}px
    ${({theme}) => theme.spacing.tall}px;
  padding: ${({theme}) => theme.spacing.tall}px
    ${({theme}) => theme.spacing.great}px;
  border-radius: ${({theme}) => theme.spacing.great}px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

export const Category = styled.View`
  margin-right: ${({theme}) => theme.spacing.great}px;
`;
export const Description = styled.View`
  flex: 1;
`;

export const BaseText = styled(Text)`
  color: ${({theme}) => theme.colors.paragraph};
  font-size: ${({theme}) => theme.fontSizes.small};
`;

export const Title = styled(BaseText)`
  font-size: ${({theme}) => theme.fontSizes.medium};
`;
export const SubTitle = styled(BaseText)``;

export const Value = styled(BaseText)<{isExpense: boolean}>`
  color: ${({theme, isExpense = true}) =>
    isExpense ? theme.colors.danger : theme.colors.success};
  font-weight: ${({theme}) => theme.fontWeight.bold};
`;
