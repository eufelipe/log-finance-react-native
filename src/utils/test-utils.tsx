/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {render, RenderAPI, RenderOptions} from '@testing-library/react-native';

import {theme} from 'styles/theme';
import {ThemeProvider} from 'styled-components/native';

import {NavigationContainer} from '@react-navigation/native';

interface Props {
  children: JSX.Element;
}

const AllTheProviders = ({children}: Props) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions,
): RenderAPI => render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
