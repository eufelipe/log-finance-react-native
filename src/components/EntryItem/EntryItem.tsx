import React from 'react';
import Swipeable from 'react-native-gesture-handler/Swipeable';

import {CategoryIcon, Currency, SwipeRemoveButton} from 'components';
import {IEntry} from 'interfaces';

import {
  Container,
  Category,
  Description,
  Title,
  SubTitle,
  Value,
} from './styles';
import {useTranslation} from 'react-i18next';

interface Props {
  entry: IEntry;
  onPressEntry: (entry: IEntry) => void;
  onPressRemoveEntry: (entry: IEntry) => void;
}

const EntryItem = ({
  entry,
  onPressEntry,
  onPressRemoveEntry,
}: Props): JSX.Element => {
  const {description, value, category, type} = entry;

  const {t} = useTranslation('entry');

  return (
    <Swipeable
      renderRightActions={() => (
        <SwipeRemoveButton onPressRemove={() => onPressRemoveEntry(entry)} />
      )}
    >
      <Container
        onPress={() => onPressEntry(entry)}
        testID="entry-item-button"
        accessible={true}
        accessibilityLabel={t('accessibility-button')}
        accessibilityHint={t('accessibility-button-hint')}
      >
        <Category>
          <CategoryIcon name={category.icon} />
        </Category>

        <Description>
          {description && <Title>{description}</Title>}
          <SubTitle>{category.description}</SubTitle>
        </Description>

        <Currency
          value={value}
          render={text => <Value isExpense={type === 'expense'}>{text}</Value>}
        />
      </Container>
    </Swipeable>
  );
};

export default EntryItem;
