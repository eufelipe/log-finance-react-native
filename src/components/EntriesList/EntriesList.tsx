import React, {useCallback} from 'react';
import {useNavigation} from '@react-navigation/native';

import {IEntry} from 'interfaces';

import EntryItem from 'components/EntryItem';
import {AddEntryNavigationProp} from 'routes/StacksRoute';

import {useEntry} from 'hooks/useEntry';
import {FlatList} from './styles';
interface Props {
  entries: IEntry[];
}

const EntriesList = ({entries}: Props): JSX.Element => {
  const navigation = useNavigation<AddEntryNavigationProp>();
  const {removeEntry} = useEntry();

  const onPressEntry = useCallback(
    (entry?: IEntry) => {
      navigation.navigate('AddEntryStack', {
        screen: 'AddEntry',
        params: {entry},
      });
    },
    [navigation],
  );

  const onPressRemoveEntry = useCallback(
    (entry: IEntry) => {
      removeEntry(entry);
    },
    [removeEntry],
  );

  const ListKeyExtractor = useCallback(item => item.id, []);

  const renderItem = useCallback(
    ({item}) => (
      <EntryItem
        entry={item}
        onPressEntry={() => onPressEntry(item)}
        onPressRemoveEntry={onPressRemoveEntry}
      />
    ),
    [onPressRemoveEntry, onPressEntry],
  );

  return (
    <FlatList
      testID="entries-list"
      data={entries}
      renderItem={renderItem}
      keyExtractor={ListKeyExtractor}
    />
  );
};

export default EntriesList;
