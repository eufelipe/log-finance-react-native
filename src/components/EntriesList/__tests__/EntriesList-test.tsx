/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import {act} from '@testing-library/react-native';

import {render} from 'utils/test-utils';
import {ICategory, IEntry} from 'interfaces';
import EntriesList from '../EntriesList';
import {FlatList} from 'react-native';
import EntryItem from 'components/EntryItem';

jest.mock('react-native-vector-icons/MaterialIcons', () => 'Icon');

jest.mock('react-i18next', () => ({
  useTranslation: () => ({t: (key: any) => key}),
}));

const CATEGORY_MOCK: ICategory = {
  id: 1,
  description: 'Restaurante',
  icon: 'money',
};

const MOCK_ENTRY: IEntry = {
  id: 1,
  description: 'Salário',
  type: 'earning',
  value: 99.45,
  date: new Date(),
  category: CATEGORY_MOCK,
};

const ENTRIES_MOCK = [
  {
    ...MOCK_ENTRY,
  },
  {
    ...MOCK_ENTRY,
    id: 2,
    description: 'Pizza',
  },
  {
    ...MOCK_ENTRY,
    id: 3,
    description: 'Hamburguer',
  },
];

const onPressEntry = jest.fn();

describe('EntriesList', () => {
  it('Test match snapshot EntriesList', () => {
    // given
    const props = {
      entries: ENTRIES_MOCK,
      onPressEntry,
    };

    // when
    const {toJSON} = render(<EntriesList {...props} />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  it('Tests if items render correctly', async () => {
    // given
    const props = {
      entries: ENTRIES_MOCK,
      onPressEntry,
    };

    // when
    const {getByText, findAllByTestId, UNSAFE_getAllByType} = render(
      <EntriesList {...props} />,
    );
    let entries;
    await act(async () => {
      entries = await findAllByTestId(/entry-item-button/i);
    });

    // then
    expect(entries).toHaveLength(props.entries.length);
    expect(UNSAFE_getAllByType(FlatList).length).toBe(1);
    expect(UNSAFE_getAllByType(EntryItem).length).toBe(props.entries.length);

    expect(getByText('Salário')).toBeTruthy();
    expect(getByText('Pizza')).toBeTruthy();
    expect(getByText('Hamburguer')).toBeTruthy();
    expect(() => getByText('no exist')).toThrow('Unable to find an element');
  });
});
