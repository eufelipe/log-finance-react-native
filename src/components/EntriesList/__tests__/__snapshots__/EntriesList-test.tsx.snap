// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntriesList Test match snapshot EntriesList 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "category": Object {
          "description": "Restaurante",
          "icon": "money",
          "id": 1,
        },
        "description": "Salário",
        "id": 1,
        "type": "earning",
        "value": 99.45,
      },
      Object {
        "category": Object {
          "description": "Restaurante",
          "icon": "money",
          "id": 1,
        },
        "description": "Pizza",
        "id": 2,
        "type": "earning",
        "value": 99.45,
      },
      Object {
        "category": Object {
          "description": "Restaurante",
          "icon": "money",
          "id": 1,
        },
        "description": "Hamburguer",
        "id": 3,
        "type": "earning",
        "value": 99.45,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {},
    ]
  }
  testID="entries-list"
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityHint="accessibility-button-hint"
        accessibilityLabel="accessibility-button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 21.428571428571427,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 10.714285714285714,
            "marginLeft": 42.857142857142854,
            "marginRight": 42.857142857142854,
            "marginTop": 10.714285714285714,
            "opacity": 1,
            "paddingBottom": 42.857142857142854,
            "paddingLeft": 21.428571428571427,
            "paddingRight": 21.428571428571427,
            "paddingTop": 42.857142857142854,
          }
        }
        testID="entry-item-button"
      >
        <View
          style={
            Array [
              Object {
                "marginRight": 21.428571428571427,
              },
            ]
          }
        >
          <Icon
            color="#464E5F"
            name="list"
            size={24}
            style={
              Array [
                Object {},
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 36,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Salário
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Restaurante
          </Text>
        </View>
        <Text
          isExpense={false}
          style={
            Array [
              Object {
                "color": "#50806A",
                "fontFamily": "Poppins-Regular",
                "fontSize": 28,
                "fontWeight": "700",
              },
            ]
          }
        >
          R$ 99,45
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityHint="accessibility-button-hint"
        accessibilityLabel="accessibility-button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 21.428571428571427,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 10.714285714285714,
            "marginLeft": 42.857142857142854,
            "marginRight": 42.857142857142854,
            "marginTop": 10.714285714285714,
            "opacity": 1,
            "paddingBottom": 42.857142857142854,
            "paddingLeft": 21.428571428571427,
            "paddingRight": 21.428571428571427,
            "paddingTop": 42.857142857142854,
          }
        }
        testID="entry-item-button"
      >
        <View
          style={
            Array [
              Object {
                "marginRight": 21.428571428571427,
              },
            ]
          }
        >
          <Icon
            color="#464E5F"
            name="list"
            size={24}
            style={
              Array [
                Object {},
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 36,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Pizza
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Restaurante
          </Text>
        </View>
        <Text
          isExpense={false}
          style={
            Array [
              Object {
                "color": "#50806A",
                "fontFamily": "Poppins-Regular",
                "fontSize": 28,
                "fontWeight": "700",
              },
            ]
          }
        >
          R$ 99,45
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityHint="accessibility-button-hint"
        accessibilityLabel="accessibility-button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 21.428571428571427,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 10.714285714285714,
            "marginLeft": 42.857142857142854,
            "marginRight": 42.857142857142854,
            "marginTop": 10.714285714285714,
            "opacity": 1,
            "paddingBottom": 42.857142857142854,
            "paddingLeft": 21.428571428571427,
            "paddingRight": 21.428571428571427,
            "paddingTop": 42.857142857142854,
          }
        }
        testID="entry-item-button"
      >
        <View
          style={
            Array [
              Object {
                "marginRight": 21.428571428571427,
              },
            ]
          }
        >
          <Icon
            color="#464E5F"
            name="list"
            size={24}
            style={
              Array [
                Object {},
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 36,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Hamburguer
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#808080",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "400",
                },
              ]
            }
          >
            Restaurante
          </Text>
        </View>
        <Text
          isExpense={false}
          style={
            Array [
              Object {
                "color": "#50806A",
                "fontFamily": "Poppins-Regular",
                "fontSize": 28,
                "fontWeight": "700",
              },
            ]
          }
        >
          R$ 99,45
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
